### PR TITLE
Delete metadata rows [BA-5864]

### PR DESCRIPTION
**Pull request by aednichols**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -128,6 +128,8 @@ lazy val drsFileSystem = (project in file("filesystems/drs"))
 
 lazy val databaseSql = (project in file("database/sql"))
   .withLibrarySettings("cromwell-database-sql", databaseSqlDependencies)
+  .dependsOn(core)
```
Comment (kshakir) [inline build.sbt]: Code smell 👃...  `core` is way too big/tempting of a large surface area of "stuff", usually associated with logic / data objects outside of SQL mappings. The `database` layer is supposed to stay as-thin-as-possible, only translating scala to slick/SQL, and leaving the logic to run in the service layer.

Comments continue below...

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 2]
```diff
@@ -358,4 +362,51 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
       labelAndKeyLabelValues, labelOrKeyLabelValues, excludeLabelAndValues, excludeLabelOrValues, submissionTimestampOption, startTimestampOption, endTimestampOption, metadataArchiveStatus, includeSubworkflows)
     runTransaction(action)
   }
+
+  override def deleteNonLabelMetadataForWorkflow(rootWorkflowId: String)(implicit ec: ExecutionContext): Future[Int] = {
+
+    val delete: DBIO[Int] = checkDeletionPreconditions(rootWorkflowId) flatMap { checkResult: Checked[Unit] =>
```
Comment (kshakir) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Again, the `database` is supposed to only contain scala-to-sql translation-- with little to zero business-logic and error-handling.

There are two directions we could go here:
- Move this logic check up into the services layer (there are examples of passing in functions, but those are for more complex cases), or
- Preferred and simpler, move this check to run down _inside the database_

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 3]
```diff
@@ -358,4 +362,51 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
...
+        s"""[Carbonite metadata deletion] Failed with non-existent root workflow "$rootWorkflowId"""".invalidNelCheck
+    }
+  }
+
+  private def isTerminalCheck(summaryEntry: WorkflowMetadataSummaryEntry): Checked[Unit] = {
+    import cromwell.core.WorkflowState
+
+    summaryEntry.workflowStatus match {
+      case Some(status) =>
+        if (WorkflowState.WorkflowStateValues.filter(_.isTerminal).map(_.toString).contains(status))
```
Comment (kshakir) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: For example, the terminal workflow statuses should be passed in as a `statusesOkToDelete: Iterable[String]` that can then be matched in the SQL / slick.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 4]
```diff
@@ -358,4 +362,51 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
...
+          ().validNelCheck
+        else
+          s"""[Carbonite metadata deletion] Failed with non-terminal summary status "$status" for workflow "${summaryEntry.workflowExecutionUuid}"""".invalidNelCheck
+      case None =>
+        s"""[Carbonite metadata deletion] Failed because summary status unexpectedly empty for workflow "${summaryEntry.workflowExecutionUuid}"""".invalidNelCheck
+    }
+  }
+
+  private def isRootWorkflowCheck(summaryEntry: WorkflowMetadataSummaryEntry): Checked[Unit] = {
+    if (summaryEntry.parentWorkflowExecutionUuid.isEmpty && summaryEntry.rootWorkflowExecutionUuid.isEmpty)
```
Comment (kshakir) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Similarly, the SQL on the server, not the scala running in the JVM, should only delete rows where the parent and the root are null.

File: database/sql/src/main/scala/cromwell/database/sql/MetadataSqlDatabase.scala [hunk 5]
```diff
@@ -139,4 +139,6 @@ trait MetadataSqlDatabase extends SqlDatabase {
                              metadataArchiveStatus: Set[Option[String]],
                              includeSubworkflows: Boolean)
                              (implicit ec: ExecutionContext): Future[Int]
+
+  def deleteNonLabelMetadataForWorkflow(rootWorkflowId: String)(implicit ec: ExecutionContext): Future[Int]
```
Comment (kshakir) [inline database/sql/src/main/scala/cromwell/database/sql/MetadataSqlDatabase.scala]: The signature here would additionally take in the `statusesOkToDelete: Iterable[String]`

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 6]
```diff
@@ -371,4 +371,5 @@ trait WorkflowMetadataSummaryEntryComponent {
       WorkflowMetadataSummaryEntry(r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<)
     })
   }
+
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: 🔙

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 2 above)
Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: >Preferred and simpler, move this check to run down inside the database

Will give this a go, should be relatively sane with the test suite in place. Thanks!

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 4 above)
Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: I think this comment may not account for the fact that the metadata entries themselves do not contain root/parent workflow IDs in a usable way (i.e. they are metadata values with no index)?

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataEntryComponent.scala [hunk 7]
```diff
@@ -62,6 +62,21 @@ trait MetadataEntryComponent {
...
+  val metadataEntriesWithoutLabelsForRootWorkflowId = Compiled(
+    (rootWorkflowId: Rep[String]) => {
+      val targetWorkflowIds = for {
+        summary <- workflowMetadataSummaryEntries
+        if summary.rootWorkflowExecutionUuid === rootWorkflowId || summary.workflowExecutionUuid === rootWorkflowId
+      } yield summary.workflowExecutionUuid
+
+      for {
+        metadata <- metadataEntries
+        if metadata.workflowExecutionUuid in targetWorkflowIds // Should use `METADATA_WORKFLOW_IDX`
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/MetadataEntryComponent.scala]: Are we sure it does?

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 8]
```diff
@@ -91,6 +91,15 @@ trait WorkflowMetadataSummaryEntryComponent {
     } yield workflowMetadataSummaryEntry.workflowStatus
   )
 
+  val isRootWorkflow = Compiled(
+    (workflowId: Rep[String]) => for {
+      summary <- workflowMetadataSummaryEntries
+      if summary.workflowExecutionUuid === workflowId
+    } yield {
+      !summary.rootWorkflowExecutionUuid.isDefined && !summary.parentWorkflowExecutionUuid.isDefined
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]:
```suggestion
      summary.rootWorkflowExecutionUuid.isEmpty && summary.parentWorkflowExecutionUuid.isEmpty
```

File: services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala [hunk 9]
```diff
@@ -304,4 +305,23 @@ trait MetadataDatabaseAccess {
...
+        Future.failed(new Exception(s"""Metadata deletion precondition failed: workflow ID "$rootWorkflowId" not found in summary table"""))
+      case (Some(false), _) =>
+        Future.failed(new Exception(s"""Metadata deletion precondition failed: workflow ID "$rootWorkflowId" is not a root workflow"""))
+      case (_, None) =>
+        Future.failed(new Exception(s"""Metadata deletion precondition failed: workflow ID "$rootWorkflowId" did not have a status in the summary table"""))
+      case (Some(true), Some(status)) =>
+        if (WorkflowState.WorkflowStateValues.filter(_.isTerminal).map(_.toString).contains(status))
+          metadataDatabaseInterface.deleteNonLabelMetadataForWorkflow(rootWorkflowId.toString)
+        else
+          Future.failed(new Exception(s"""Metadata deletion precondition failed: workflow ID "$rootWorkflowId" was in non-terminal status "$status""""))
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala]: Squawking Out Loud:
In `WorkflowState`:

```
  lazy val TerminalWorkflowStateValues = WorkflowState.WorkflowStateValues.filter(_.isTerminal)

  def isTerminal(stringState: String): Boolean = {
    WorkflowState.TerminalWorkflowStateValues.map(_.toString).contains(stringState)
  }
```

Comment (mcovarr) [reply services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala]: Here:
```
      case (Some(true), Some(state)) if WorkflowState.isTerminal(state) =>
        metadataDatabaseInterface.deleteNonLabelMetadataForWorkflow(rootWorkflowId.toString)
      case (Some(true), Some(state)) =>
        Future.failed(new Exception(s"""Metadata deletion precondition failed: workflow ID "$rootWorkflowId" was in non-terminal status "$state""""))
```

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataEntryComponent.scala (hunk 7 above)
Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/tables/MetadataEntryComponent.scala]: That is a great question, I should `EXPLAIN`

File: services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala (hunk 9 above)
Comment (aednichols) [reply services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala]: Looking at `WorkflowState` gave me an even better idea:
```
WorkflowState.withName(status).isTerminal
```

**Comment by aednichols** [comment]
An `EXPLAIN` of the generated SQL
```
delete `METADATA_ENTRY` 
from 
  `METADATA_ENTRY` 
where 
  (
    `METADATA_ENTRY`.`WORKFLOW_EXECUTION_UUID` in (
      select 
        `WORKFLOW_EXECUTION_UUID` 
      from 
        `WORKFLOW_METADATA_SUMMARY_ENTRY` 
      where 
        (
          `ROOT_WORKFLOW_EXECUTION_UUID` = 'workflow id: 3 to delete, 1 label'
        ) 
        or (
          `WORKFLOW_EXECUTION_UUID` = 'workflow id: 3 to delete, 1 label'
        )
    )
  ) 
  and (
    not (
      `METADATA_ENTRY`.`METADATA_KEY` like 'labels:%'
    )
  )
```
identifies `METADATA_WORKFLOW_IDX` in `possible_keys`.

On a local DB of trivial size, the index isn't actually used, checking a more realistic scenario on an alpha clone.

**Comment by aednichols** [comment]
Alpha test looks good, plan is using `METADATA_WORKFLOW_IDX` on `METADATA_ENTRY`.

Actually deleting rows (on an alpha clone) completes in 100-200ms for small workflows.

This is an `EXPLAIN` in IntelliJ, exported as an HTML table – apparently you can just paste HTML into Github and it works! 

<!DOCTYPE html>
<html>
<body>
<table border="1" style="border-collapse:collapse">
<tr><th>id</th><th>select_type</th><th>table</th><th>type</th><th>possible_keys</th><th>key</th><th>key_len</th><th>ref</th><th>rows</th><th>Extra</th></tr>
<tr><td>1</td><td>SIMPLE</td><td>WORKFLOW_METADATA_SUMMARY_ENTRY</td><td>index_merge</td><td>UC_WORKFLOW_METADATA_SUMMARY_ENTRY_WEU,IX_WORKFLOW_METADATA_SUMMARY_ENTRY_RWEU</td><td>IX_WORKFLOW_METADATA_SUMMARY_ENTRY_RWEU,UC_WORKFLOW_METADATA_SUMMARY_ENTRY_WEU</td><td>303,302</td><td>NULL</td><td>2</td><td>Using union(IX_WORKFLOW_METADATA_SUMMARY_ENTRY_RWEU,UC_WORKFLOW_METADATA_SUMMARY_ENTRY_WEU); Using where</td></tr>
<tr><td>1</td><td>SIMPLE</td><td>METADATA_ENTRY</td><td>ref</td><td>METADATA_WORKFLOW_IDX</td><td>METADATA_WORKFLOW_IDX</td><td>767</td><td>cromwell.WORKFLOW_METADATA_SUMMARY_ENTRY.WORKFLOW_EXECUTION_UUID</td><td>73</td><td>Using where</td></tr></table>
</body>
</html>

**Pull request merged**